eventuals/collect.h

```diff
@@ -53,20 +53,99 @@ struct EventualsCollector<
 
 ////////////////////////////////////////////////////////////////////////
 
-// TODO(folming): Issue #486.
+struct _Collect final {
+  template <typename K_, typename Collection_>
+  struct Continuation final {
+    Continuation(K_ k)
+      : k_(std::move(k)) {}
+
+    void Begin(TypeErasedStream& stream) {
+      stream_ = &stream;
+      stream_->Next();
+    }
+
+    template <typename Error>
+    void Fail(Error&& error) {
+      k_.Fail(std::forward<Error>(error));
+    }
+
+    void Stop() {
+      k_.Stop();
+    }
+
+    template <typename Arg>
+    void Body(Arg&& value) {
+      EventualsCollector<Collection_>::Collect(
+          collection_,
+          std::forward<Arg>(value));
+      stream_->Next();
+    }
+
+
+    void Ended() {
+      k_.Start(std::move(collection_));
+    }
+
+    void Register(Interrupt& interrupt) {
+      CHECK(interrupt_ == nullptr);
+      interrupt_ = &interrupt;
+      k_.Register(interrupt);
+    }
+
+    Collection_ collection_ = Collection_();
+
+    TypeErasedStream* stream_ = nullptr;
+
+    Interrupt* interrupt_ = nullptr;
+
+    // NOTE: we store 'k_' as the _last_ member so it will be
+    // destructed _first_ and thus we won't have any use-after-delete
+    // issues during destruction of 'k_' if it holds any references or
+    // pointers to any (or within any) of the above members.
+    K_ k_;
+  };
+
+  template <template <typename...> class Collection_>
+  struct Composable final {
+    template <typename Arg>
+    using ValueFrom = Collection_<std::decay_t<Arg>>;
+
+    template <typename Arg, typename Errors>
+    using ErrorsFrom = Errors;
+
+    template <typename Arg, typename K>
+    auto k(K k) && {
+      return Continuation<K, Collection_<std::decay_t<Arg>>>(std::move(k));
+    }
+  };
+
+  template <typename Collection_>
+  struct ComposableTyped final {
+    template <typename Arg>
+    using ValueFrom = Collection_;
+
+    template <typename Arg, typename Errors>
+    using ErrorsFrom = Errors;
+
+    template <typename Arg, typename K>
+    auto k(K k) && {
+      return Continuation<K, Collection_>(std::move(k));
+    }
+  };
+};
+
+////////////////////////////////////////////////////////////////////////
+
+template <template <typename...> class Collection>
+[[nodiscard]] auto Collect() {
+  return _Collect::Composable<Collection>();
+}
```

Comment:
What if we made the "typed" version just return what we always had:

```cpp
    return Loop<Collection>()
      .context(Collection())
      .body([](auto& collection, auto& stream, auto&& value) {
        EventualsCollector<Collection>::Collect(
            collection,
            std::forward<decltype(value)>(value));
        stream.Next();
      })
      .ended([](auto& collection, auto& k) {
        k.Start(std::move(collection));
      });
```

Then we made this version be defined _after_ and just do:

```suggestion
struct _Collect {
  template <template <typename...> class Collection_>
  struct Composable {
    template <typename Arg>
    using ValueFrom = Collection_<std::decay_t<Arg>>;

    template <typename Arg, typename Errors>
    using ErrorsFrom = Errors;

    template <typename Arg, typename K>
    auto k(K k) && {
      return Build(Collect<Collection_<std::decay_t<Arg>>>(), std::move(k));
    }  
  };
};

template <template <typename...> class Collection>
[[nodiscard]] auto Collect() {
  return _Collect::Composable<Collection>{};
}
```

As we "grow" up as a library we strive to only use our existing abstractions! When we need to create a "new' abstraction it always begs to ask if our existing ones are not powerful enough. If you see a reason why this won't work let me know and I'll review `_Collect::Continuation`!

Reply:
And if you can think of an abstraction that lets us do this all without ever creating the `_Collect` struct that would be even better!